[Docs] Security Route Configuration (#193994)
elastic/kibana: dev_docs/key_concepts/api_authorization.mdx

---
id: kibDevDocsSecurityAPIAuthorization
slug: /kibana-dev-docs/key-concepts/security-api-authorization
title: Kibana API authorization
description: This guide provides an overview of API authorization in Kibana.
date: 2024-10-04
tags: ['kibana', 'dev', 'contributor', 'security']
---

> Security is everyone's responsibility. This is inclusive of design, product, and engineering. The purpose of this guide is to give a high-level overview of security constructs and expectations.

Reviewer: This intro doesn't quite fit the more narrow focus of API authorization.

```suggestion
Authorization is an important aspect of API design. It must be considered for all endpoints, even those marked as `internal` . This guide explains how and when to apply authorization to your endpoints.
```

- [Versioned Router Security Configuration Examples](#versioned-router-security-configuration-examples)
- [Authorization Response in Route Handlers](#authorization-response-available-in-route-handlers)
- [OpenAPI Specification (OAS) documentation](#openapi-specification-oas-documentation)
- [Migrating from `access` Tags to `security` Configuration](#migrating-from-access-tags-to-security-configuration)

## API authorization
Kibana API routes do not have any authorization checks applied by default. This means that your APIs are accessible to anyone with valid credentials, regardless of their permissions. This includes users with no roles assigned.
This on its own is insufficient, and care must be taken to ensure that only authorized users can invoke your endpoints.

Kibana leverages <DocLink id="kibDevDocsSavedObjectsIntro" text="Saved Objects" /> for a majority of its persistence. The Saved Objects Service performs its own authorization checks, so if your API route is primarily a CRUD interface to Saved Objects, then your authorization needs are already met.
This is also true for derivatives of the Saved Objects Service, such as the Alerting and Cases services.

> If your endpoint is not a CRUD interface to Saved Objects, then your route ensure that only authorized users can invoke your endpoint. This is **especially** important if your route does any of the following:

Reviewer: I know this was copy/paste from the old docs, but I thought this was worth clarifying.

```suggestion
If your endpoint is not a CRUD interface to Saved Objects, or if your endpoint bypasses our built-in Saved Objects authorization checks, then you must ensure that only authorized users can invoke your endpoint. This is **especially** important if your route does any of the following:
```

1. Performs non-insignificant processing, causing load on the Kibana server.
2. Calls Elasticsearch using the internal `kibana_system` user.
3. Calls a third-party service.
4. Returns any non-public information to the caller, such as system configuration or state.

> ## [Deprecated] Adding API authorization with `access` tags

Reviewer: I know the section is right below this one, but can we mention that this was deprecated in favor of the new `security` configuration?

Example configuration:
```ts
router.get({
  path: '/api/path',
  options: {
    tags: ['access:<privilege_1>', 'access:<privilege_2>'],
  },
  ...
}, handler);
```

More information on adding `access` tags to your routes can be found temporarily in the [legacy documentation](https://www.elastic.co/guide/en/kibana/current/development-security.html#development-plugin-feature-registration)

### Why not add `access` tags to all routes by default?
Each authorization check that we perform involves a round-trip to Elasticsearch, so they are not as cheap as we'd like. Therefore, we want to keep the number of authorization checks we perform within a single route to a minimum.
Adding an `access` tag to routes that leverage the Saved Objects Service would be redundant in most cases, since the Saved Objects Service will be performing authorization anyway.

## Adding API authorization with `security` configuration
`KibanaRouteOptions` provides a security configuration at the route definition level, offering robust security configurations for both **Classic** and **Versioned** routes.

> ### Key features of the new security configuration:

Reviewer: This is currently "new", but I know that I'll personally forget to remove this once it's no longer new. Let's shorten the heading to keep this future-proof.

```suggestion
### Key features:
```

4. **Improved Documentation with OpenAPI (OAS)**: Automatically generated OAS documentation with the required privileges for each route.
5. **AuthzResult Object in Route Handlers**: Access the authorization response in route handlers to see which privileges were met.

### Configuring Authorization on Routes
**Before Migration:**
```ts
router.get({
  path: '/api/path',
  options: {
    tags: ['access:<privilege_1>', 'access:<privilege_2>'],
  },
  ...
}, handler);
```

**After Migration:**
```ts
router.get({
  path: '/api/path',
  security: {
    authz: {
      requiredPrivileges: ['<privilege_1>', '<privilege_2>'],
    },
  },
  ...
}, handler);
```

### Opting Out of Authorization for Specific Routes
**Before Migration:**
```ts
router.get({
  path: '/api/path',
  ...
}, handler);
```

**After Migration:**
```ts
router.get({
  path: '/api/path',
  security: {
    authz: {
      enabled: false,
      reason: 'This route is opted out from authorization because ...',
    },
  },
  ...
}, handler);
```

### Classic Router Security Configuration Examples

**Example 1: All Privileges Required**

> Requires `privilege_1` AND both `privilege_2` to access the route.

Reviewer: ```suggestion
Requires `privilege_1` AND `privilege_2` to access the route.
```

```ts
router.get({
  path: '/api/path',
  security: {
    authz: {
      requiredPrivileges: ['<privilege_1>', '<privilege_2>'],
    },
  },
  ...
}, handler);
```

**Example 2: Any Privileges Required**
Requires `privilege_1` OR `privilege_2` to access the route.
```ts
router.get({
  path: '/api/path',
  security: {
    authz: {
      requiredPrivileges: [{ anyRequired: ['<privilege_1>', '<privilege_2>'] }],
    },
  },
  ...
}, handler);
```

**Example 3: Complex Configuration**
Requires `privilege_1` AND `privilege_2` AND (`privilege_3` OR `privilege_4`) to access the route.
```ts
router.get({
  path: '/api/path',
  security: {
    authz: {
      requiredPrivileges: [{ allRequired: ['<privilege_1>', '<privilege_2>'], anyRequired: ['<privilege_3>', '<privilege_4>'] }],
    },
  },
  ...
}, handler);
```

### Versioned Router Security Configuration Examples

> Versioned router allows different security configurations registered per version, giving flexibility when evolving API security over time.

Reviewer: ```suggestion
Different security configurations can be applied to each version when using the Versioned Router. This allows your authorization needs to evolve in lockstep with your API.
```

@azasypkin I think it'd be nice to include a link here to privilege deprecation documentation (once you have it available). I don't want to give engineers the wrong idea here. I worry this could be easily misunderstood, and lead to insufficient authorization checks as new API versions are introduced.

2. **Version 1**: Requires **both** `<privilege_1>` and `<privilege_2>` privileges.

3. **Version 2**: Requires `<privilege_3>` AND (`<privilege_1>` OR `<privilege_2>`).

4. **Version 3**: Requires only `<privilege_3>`.

```ts
router.versioned
  .get({
    path: '/internal/path',
    access: 'internal',
    // default security configuration, will be used for version unless overridden
    security: {
      authz: {
        requiredPrivileges: ['<privilege_1>'],
      },
    },
  })
  .addVersion({
    version: '1',
    validate: false,
    security: {
      authz: {
        requiredPrivileges: ['<privilege_1>', '<privilege_2>'],
      },
    },
  }, handlerV1)
  .addVersion({
    version: '2',
    validate: false,
    security: {
      authz: {
        requiredPrivileges: ['<privilege_3>', anyRequired: ['<privilege_1>', '<privilege_2>']],
      },
    },
  }, handlerV2)
  .addVersion({
    version: '3',
    validate: false,
    security: {
      authz: {
        requiredPrivileges: ['<privilege_3>'],
      },
    },
  }, handlerV3);
```

## Authorization Response Available in Route Handlers

> The `AuthzResult` object is available in route handlers, which provides information about the authorization response.

Reviewer: ```suggestion
The `AuthzResult` object is available in route handlers, which provides information about the privileges granted to the caller.
```

> For example, you have a route that requires  `privilege_3` and ANY of the privileges `privilege_1` OR `privilege_2`:

Reviewer: consistency with brackets
```suggestion
For example, you have a route that requires  `<privilege_3>` and ANY of the privileges `<privilege_1>` OR `<privilege_2>`:
```

```ts
router.get({
  path: '/api/path',
  security: {
    authz: {

>       requiredPrivileges: ['privilege_3', { anyRequired: ['<privilege_1>', '<privilege_2>'] }],

Reviewer: ```suggestion
      requiredPrivileges: ['<privilege_3>', { anyRequired: ['<privilege_1>', '<privilege_2>'] }],
```

    },
  },
  ...
}, (context, request, response) => {
  // The authorization response is available in `request.authzResult`
  // {

>   //   "privilege_3": true,
>   //   "privilege_1": true,
>   //   "privilege_2": false

Reviewer: ```suggestion
  //   "<privilege_3>": true,
  //   "<privilege_1>": true,
  //   "<privilege_2>": false
```

  // }
});
```

## OpenAPI Specification (OAS) documentation
Based on the security configuration defined in routes, OAS documentation will automatically generate and include description about the required privileges. 

> This makes it easy to view the security requirements of each endpoint in a standardized format, facilitating better understanding and usage by developers or teams consuming the API

Reviewer: ```suggestion
This makes it easy to view the security requirements of each endpoint in a standardized format, facilitating better understanding and usage by developers or teams consuming the API.
```

To check the OAS documentation for a specific API route and see its security details, you can use the following command:
```sh
GET /api/oas?pathStartsWith=/your/api/path
```

## Migrating from `access` Tags to `security` Configuration
We aim to use the same privileges that are currently defined with tags `access:<privilege_name>`.
To assist with this migration, we have created eslint rule `no_deprecated_authz_config`, that will automatically convert your `access` tags to the new `security` configuration.
It scans route definitions and converts `access` tags to the new `requiredPriviliges` configuration.

Note: The rule is disabled by default to avoid automatic migrations without an oversight. You can perform migrations by running:

**Migrate Routes with Defined Authorization**
```sh
MIGRATE_DISABLED_AUTHZ=false npx eslint --ext .ts --fix --rule "@kbn/eslint/no_deprecated_authz_config:error" path/to/your/folder
```
We encourage you to migrate routes that are opted out from authorization to new config and provide legitimate reason for disabled authorization.
It is better to migrate routes opted out from authorization iteratively and elaborate on the reasoning. 

> Probably, you can consider enabling authorization for these routes in the future.

Reviewer: ```suggestion
Routes without a compelling reason to opt-out of authorization should plan to introduce them as soon as possible.
```

**Migrate All Routes (Including Those Opted Out from Authorization)**
```sh
MIGRATE_DISABLED_AUTHZ=true npx eslint --ext .ts --fix --rule "@kbn/eslint/no_deprecated_authz_config:error" path/to/your/folder
```

## Questions?

> If you have any questions or need help with API authorization, please reach out to the Kibana Security team on the `#kibana-security` Slack channel.

Reviewer: These docs are in the public repo, so it may be best to instead mention our GH team.
```suggestion
If you have any questions or need help with API authorization, please reach out to the `@elastic/kibana-security` team.
```

---
id: kibDevDocsSecurityAPIAuthorization
slug: /kibana-dev-docs/key-concepts/security-api-authorization
title: Kibana API authorization
description: This guide provides an overview of API authorization in Kibana.
date: 2024-10-04
tags: ['kibana', 'dev', 'contributor', 'security']
---

Authorization is an important aspect of API design. It must be considered for all endpoints, even those marked as `internal` . This guide explains how and when to apply authorization to your endpoints

> This guide specifically focuses on API authorization:

Author: > just a list of the most important
I wanted to highlight the most important, but if it's confusing I can make it as table of contents

> I'm wondering if we can just use normal case and use capital letters only where necessary, but it's minor.
I'm happy to convert it

> This guide specifically focuses on API authorization:

Reviewer 2: > I wanted to highlight the most important, but if it's confusing I can make it as table of contents

Got it, no strong preference - it might just be me. Whatever you feel is best, would work for me too.

> I'm happy to convert it

That'd be great, thank you!
